lab5/src/ll1/Parser.java

```diff
@@ -0,0 +1,136 @@
+                    configuration.pop();
+                } else {
+                    if (Objects.equals(parsingTableCell, ParsingTableCell.getAcceptCell())) {
+                        go = false;
+                        accepted = true;
+                    } else {
+                        System.out.println("ParsingTable error at: "+ configuration.getHeadOfWorkingStack() + ',' + configuration.getHeadOfInputStack());
+                        go = false;
+                    }
+                }
+            }
+        }
+        if (accepted) {
+            return configuration.getOutputStack();
+        }
+        return null;
+    }
+
+    public List<TreeNode> getTree(List<String> sequence) {
+        List<Integer> sequenceOfProductions = getSequenceOfProductions(sequence);
+        return convertSequenceOfProductionsIntoTree(sequenceOfProductions);
+    }
+
+    public List<TreeNode> convertSequenceOfProductionsIntoTree(List<Integer> sequenceOfProductions) {
+        List<TreeNode> treeMatrix = new ArrayList<>();
+        treeMatrix.add(new TreeNode(1, grammar.getStartingSymbol(), null, new LinkedList<>()));
+        for (Integer indexOfProduction : sequenceOfProductions) {
+            TreeNode parentNode = getLeftMostNonTerminalNodeFromTree(treeMatrix.get(0));
+            Production production = grammar.getProductionWithIndex(indexOfProduction);
+            List<String> productionRightSide = production.getRightSide();
+            int index = treeMatrix.size() + 1; //new nodes will start with this index
+            for (int childCount = 0; childCount < productionRightSide.size(); childCount++) {
+                TreeNode newChild;
+                String info = productionRightSide.get(childCount);
+                if (grammar.getNonTerminals().contains(info)) {
+                    //if it is a non-terminal we'll put an empty list for children
+                    newChild = new TreeNode(index + childCount, productionRightSide.get(childCount), parentNode,
+                            new LinkedList<>());
+                } else {
+                    //if it is a terminal we put null.
+                    newChild = new TreeNode(index + childCount, productionRightSide.get(childCount), parentNode, null);
+                }
+                treeMatrix.add(newChild);
+                parentNode.getChildren().add(newChild);
+            }
+        }
+        writeToFile(treeMatrix);
+        return treeMatrix;
+    }
+
+    private boolean isSpecialCase(ParsingTableCell parsingTableCell) {
+        return Objects.equals(parsingTableCell, ParsingTableCell.getAcceptCell())
+                || Objects.equals(parsingTableCell, ParsingTableCell.getPopCell())
+                || Objects.equals(parsingTableCell, ParsingTableCell.getErrorCell());
+    }
+
+    private TreeNode getLeftMostNonTerminalNodeFromTree(TreeNode currentNode) {
+        if (currentNode != null) { //this is dfs basically
```

Comment:
You should rethink this method. It looks too nested, but it can be simplified. Check this out -> https://www.youtube.com/watch?v=CFRhGnuXG-4